float32 radius
float32 theta
float32 phi
